### PR TITLE
Keithley2231A DC Power Supply

### PR DESCRIPTION
**Pull request by marcogobbo**
Checklist before review:
- [x] Test on hardware.

Checklist before merge:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Documentation is updated.

### CODE REVIEW

File: src/qinst/serial/keithley2231a.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        self.write(f"*RCL {self.validate_range(memory, 0, 30)}")
+
+    def wait(self):
+        """Prevent the instrument from executing commands until all commands are completed."""
+        self.write("*WAI")
+
+    @property
+    def ch_select(self) -> str:
+        """Select the channel to use."""
+        return self.query("INST:NSEL?")
```
Comment (PietroCampana) [inline src/qinst/serial/keithley2231a.py:49]:
```suggestion
    def ch_select(self) -> int:
        """Select the channel to use."""
        return int(self.query("INST:NSEL?"))
```

File: src/qinst/serial/keithley2231a.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    def voltage(self) -> float:
+        """Voltage of the selected the channel."""
+        return self.query("SOUR:VOLT:LEV:IMM:AMPL?")
+
+    @voltage.setter
+    def voltage(self, ch: int, value: float):
+        self.ch_select = self.validate_opt(ch, (1, 2, 3))
+        self.write(
+            f"SOUR:VOLT:LEV:IMM:AMPL {self.validate_range(value, 0, 5) if ch == 3 else self.validate_range(value, 0, 30)}"
+        )
```
Comment (PietroCampana) [inline src/qinst/serial/keithley2231a.py:65]:
```suggestion
    def voltage(self, value: float):
        v_max = 5 if self.ch_select == 3 else 30
        self.write(
            f"SOUR:VOLT:LEV:IMM:AMPL {self.validate_range(value, 0, vmax)}"
        )
```

File: src/qinst/serial/keithley2231a.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        return self.query("INST:NSEL?")
+
+    @ch_select.setter
+    def ch_select(self, ch: int):
+        self.write(f"INST:NSEL {self.validate_opt(ch, (1,2,3))}")
+
+    @property
+    def voltage(self) -> float:
+        """Voltage of the selected the channel."""
+        return self.query("SOUR:VOLT:LEV:IMM:AMPL?")
```
Comment (PietroCampana) [inline src/qinst/serial/keithley2231a.py:58]:
```suggestion
        return float(self.query("SOUR:VOLT:LEV:IMM:AMPL?"))
```

File: src/qinst/serial/keithley2231a.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    def voltage(self, ch: int, value: float):
+        self.ch_select = self.validate_opt(ch, (1, 2, 3))
+        self.write(
+            f"SOUR:VOLT:LEV:IMM:AMPL {self.validate_range(value, 0, 5) if ch == 3 else self.validate_range(value, 0, 30)}"
+        )
+
+    @property
+    def current(self) -> float:
+        """Current of the selected the channel."""
+        return self.query("SOUR:CURR:LEV:IMM:AMPL?")
```
Comment (PietroCampana) [inline src/qinst/serial/keithley2231a.py:70]: Remember to cast the return variable to the specified type

File: src/qinst/serial/keithley2231a.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    @property
+    def current_limit(self) -> float:
+        """Current limit of the selected the channel."""
+        return self.query("SOUR:CURR:LIMIT:LEV?")
+
+    @current_limit.setter
+    def current_limit(self, ch: int, value: float):
+        self.ch_select = self.validate_opt(ch, (1, 2, 3))
+        self.write(f"SOUR:VOLT:LIMIT:LEV {self.validate_range(value, 0, 3)}")
```
Comment (PietroCampana) [inline src/qinst/serial/keithley2231a.py:97]: delete "ch" argument from properties

**Review by PietroCampana — CHANGES_REQUESTED** [review]

File: src/qinst/serial/keithley2231a.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    def connect(self):
+        """Put Keithley 2231A DC Power Supply in remote."""
+        self.serial.open()
+        self.write("SYST:REM")
+
+    def disconnect(self):
+        """Take Keithley 2231A DC Power Supply out of remote."""
+        if self.serial.is_open:
+            self.write("SYST:LOC")
```
Comment (rodolfocarobene) [inline src/qinst/serial/keithley2231a.py:22]: For uniformity, add log.info for connection and disconection

**Review by PietroCampana — APPROVED** [review]

**Pull request merged**